### ISSUE TITLE
Length normalization in DPO and other variants

### ISSUE THREAD

**Issue by yakazimir** [princeton-nlp/SimPO#20]
Given how crucial length normalization seems to be in the success of SimPO, and how easy it is to change in the these implementations, did you experiment with length normalized DPO, or using length normalization with any of these other DPO variants? (it's interesting to note, as you do in the paper, that ODPO also uses length normalization)

There has been some discussion of this here (https://github.com/eric-mitchell/direct-preference-optimization/issues/48) and here (https://github.com/eric-mitchell/direct-preference-optimization/issues/40). 

I'm trying to investigate this myself, just curious if it has already been tried. Thank you. 

_Originally posted by @yakazimir in https://github.com/princeton-nlp/SimPO/issues/4#issuecomment-2140938948_

**Comment by yumeng5** [comment]
Hi,

Sorry for missing your original comment posted in the closed issue. 

We have tried DPO with length normalization, and it indeed worked better than the original DPO under our Base setups. However, it performs similarly to DPO under Instruct setups. This is probably because the original UltraFeedback preference dataset (used in the Base setups) has clear length bias (i.e., winning responses are generally longer than losing responses) whereas the Instruct setups do not -- the winning and losing responses are regenerated with instruction-tuned LMs and their lengths are usually comparable. In short, length normalization helps DPO too when there's length bias in the training set.

Compared to SimPO, I think DPO will benefit less from length normalization though, and this is because DPO already implicitly implements length normalization with the usage of the reference model (Section 4.4 in our paper). Overall, we observed that DPO + length normalization still underperforms SimPO for all setups, as the presence of the reference model still causes the discrepancy between training and decoding.

I quickly looked at the two issues you pointed out in the original DPO repo regarding length normalization. It looked like applying length normalization has not been widely recognized as helpful for DPO. We didn't apply any tricks to make length normalization work for DPO except tuning the beta hyperparameter -- For example, without length normalization, the optimal beta needs to be small (e.g., 0.01 or 0.1), whereas with length normalization, the beta value needs to be much bigger (e.g., 2.0).

I hope this is helpful!

Best,
Yu

**Comment by yakazimir** [comment]
This is very helpful indeed, thanks for taking the time to write back. 

If you have some of this analysis in a tangible form, i think adding it to a future version of your paper would really strenghten your some of your points about SimPO's improvement over DPO :).  

Closing for now. best wishes.

**Comment by Isaaclgz** [comment]
Just contributing to this discussion since I recently went down this rabbit hole too, I noticed both DPO and KTO implementations on trl default to summing log-probabilities instead of averaging. 

In my experiments applying KTO to safety alignment, averaging of log-probabilities did indeed lead to very unstable training. I think @yumeng5  suggestion of higher beta makes a lot of intuitive sense - averaging log-probabilities leads to smaller gradients, so beta must be larger (which I did not test, which could explain the ineffective training).
